docs: add search to tokens page

ninamarina opened the pull request:
# docs: add search to tokens page

<!--- Feel free to remove any unused sections -->

## Obligatory GIF (super important!)

<!--- go to giphy.com, pick a gif, click share, copy gif link, paste within the () brackets below. -->

![Obligatory GIF](https://media.giphy.com/media/JUh0yTz4h931K/giphy.gif?cid=ecf05e474dd7o3q5e2zv48vns11uusar2zyw62tnr6bzsp2m&ep=v1_gifs_search&rid=giphy.gif&ct=g)

## :hammer_and_wrench: Type Of Change

<!--- Tick or place an `x` for the type of change. Should match the type in the commit message / PR title
in https://github.com/dialpad/dialtone/blob/staging/.github/COMMIT_CONVENTION.md -->

These types will increment the version number on release:

- [ ] Fix
- [ ] Feature
- [ ] Performance Improvement
- [ ] Refactor

These types will not increment the version number, but will still deploy to documentation site on release:

- [x] Documentation
- [ ] Build system
- [ ] CI
- [ ] Style (code style changes, not css changes)
- [ ] Test
- [ ] Other (chore)

## :book: Jira Ticket
https://dialpad.atlassian.net/browse/DLT-1505
<!--- Enter the URL of the Jira ticket associated with this PR -->

## :book: Description
Adds search to the tokens page.
You can search by name, category or description.
Adds a dynamic table of contents, using vuepress toc component.
<!--- Describe specifically what the changes are -->

## :pencil: Checklist

<!--- Tick or place an `x` in all of the checkboxes that apply -->
<!--- Remove checkboxes that do not apply -->

For all PRs:

- [x] I have ensured no private Dialpad links or info are in the code or pull request description (Dialtone is a public repo!).
- [x] I have reviewed my changes.
- [x] I have added all relevant documentation.
- [x] I have considered the performance impact of my change.

For all CSS changes:

- [x] I have used design tokens whenever possible.
- [x] I have considered how this change will behave on different screen sizes.
- [x] I have visually validated my change in light and dark mode.
- [x] I have used gap or flexbox properties for layout instead of margin whenever possible.

## :crystal_ball: Next Steps
Search by keywords instead of description: https://dialpad.atlassian.net/browse/DLT-1607
<!--- Describe any future changes that need to be made after merging the PR, especially any follow up tasks after release. -->

## :camera: Screenshots / GIFs
![search-tokens](https://github.com/dialpad/dialtone/assets/24460973/f61ff377-2ad4-46f8-a045-fdc2d119bc94)


<!--- Add these if necessary. Since we have deploy previews for every PR it may not always be. -->
<!--- Link any screenshots / GIFs below -->

ninamarina committed:
refactor token table
refactor to make code more readable and to process the tokens before mount

ninamarina committed:
add search functionality

yorbi-dp commented:
Super stoked for this! 🔥

yorbi-dp reviewed (changes requested):
Thank you, Nina! This is looking great. Just a couple of comments related to the expected results when searching:

- When searching for multiple word token names, I expect to get the token in the results. For example, searching `foreground primary` should bring everything that has `foreground primary` in that order, such as `var(--dt-color-foreground-primary)`.
- When searching for full token names, I'm not getting any results. See the example.
<img width="1048" alt="Screenshot 2024-03-05 at 4 00 15 PM" src="https://github.com/dialpad/dialtone/assets/144369133/62fdba55-5502-4535-828a-b2b1b1a492e4">
- I'm still not 100% convinced about using the description to search because it's bringing results I'm not expecting. For example, searching surface also brings up foreground colors because it's in the description. Probably not for this ticket, but can we add keywords instead of using the description for each token?

These are all my comments for now. Thank you Nina!

ninamarina committed:
filter categories and add no results message

ninamarina committed:
improve search

ninamarina commented:
> Thank you, Nina! This is looking great. Just a couple of comments related to the expected results when searching:
> 
> * When searching for multiple word token names, I expect to get the token in the results. For example, searching `foreground primary` should bring everything that has `foreground primary` in that order, such as `var(--dt-color-foreground-primary)`.

Fixed in last commit:
<img width="1124" alt="image" src="https://github.com/dialpad/dialtone/assets/24460973/31159ca2-34ba-4d70-84d4-da037e8e4370">

Notice that is also brings `var(--dt-color-foreground-tertiary)` as a result because it has "primary" in the description. That would be fixed when we look for keywords instead of description

> * When searching for full token names, I'm not getting any results. See the example.

Fixed in last commit:
<img width="1073" alt="image" src="https://github.com/dialpad/dialtone/assets/24460973/83ee0b0f-a290-493d-8f21-902ba6202d91">

> - I'm still not 100% convinced about using the description to search because it's bringing results I'm not expecting. For example, searching surface also brings up foreground colors because it's in the description. Probably not for this ticket, but can we add keywords instead of using the description for each token?

I totally agree this would be better. Will look into where can we add this keywords.

yorbi-dp reviewed (changes requested):
This is looking awesome, Nina! Thanks.

One improvement: when searching for styles that are in the Figma structure, e.g., `typography/body/compact/small`, I would expect to get the token that is represented by that style, `var(--dt-typography-body-compact-small)`.
![image](https://github.com/dialpad/dialtone/assets/144369133/b2cb268e-c119-4265-a119-32da821496de)
Expected result:
![image](https://github.com/dialpad/dialtone/assets/144369133/43fbb527-5280-4199-a756-627c2dfc9d4f)

ninamarina committed:
fix to filter when changin format and theme

ninamarina committed:
search by figma name

ninamarina committed:
add dynamic toc

yorbi-dp reviewed (approved):
This is looking great Nina! Thank you.

ninamarina committed:
fix size and space tables in design pages

iropolo commented on apps/dialtone-documentation/docs/.vuepress/baseComponents/AllTokens.vue:
```diff
@@ -1,5 +1,43 @@
+      </template>
+      <template #rightIcon>
+        <dt-button
+          v-if="!isSearchEmpty"
```
`isSearchEmpty` is being used once and using `!`

Could we call it sth like:
`isSearchNotEmpty`
`const isSearchNotEmpty = computed(() => searchInput.value && searchInput.value.trim().length > 0);`

braddialpad replied:
Maybe just `hasSearchTerm`. I usually try to avoid variable names with 'Not' in them because it's confusing.

iropolo replied:
Agree, sure, it can be confusing sometimes.

iropolo commented on apps/dialtone-documentation/docs/.vuepress/baseComponents/AllTokens.vue:
```diff
@@ -14,50 +52,243 @@
-    </slot>
-    <token-table :category="category" :format="format" :theme="theme" />
+  </div>
+  <div v-for="category in Object.keys(filteredTokens)" :key="category">
```
Seems like we are using `Object.keys(filteredTokens)` in many places.

Could we create a 

`const filteredTokensKeys = computed(() => Object.keys(filteredTokens.value));`

and replace them.

iropolo commented on apps/dialtone-documentation/docs/.vuepress/baseComponents/AllTokens.vue:
```diff
@@ -14,50 +52,243 @@
+      {{ category }}
+    </h2>
+    <p class="dialtone-intro">
+      {{ categoryIntroAndDesc[category]?.intro }}
```
huh, I would like to void use inline expression in templates.
Instead of this what if we use a method like:

```
const getCategoryIntro = (category) => {
  return categoryIntroAndDesc[category]?.intro;
};
```

```
<p
      v-if="getCategoryIntro(category)"
      class="dialtone-intro">
      {{ getCategoryIntro(category) }}
    </p>
```

Then we avoid to create an empty <p> element.

iropolo commented on apps/dialtone-documentation/docs/.vuepress/baseComponents/AllTokens.vue:
```diff
@@ -14,50 +52,243 @@
+    <p class="dialtone-intro">
+      {{ categoryIntroAndDesc[category]?.intro }}
+    </p>
+    <div v-html="categoryIntroAndDesc[category]?.description" />
```
Here also is creating an empty `<div>` and it has an inline expression, we can replicate what we did previously with a method.

We are calling these method twice, we can avoid it with a computed property but since this is not reactive I dont see it necessary, and the method is pretty straightforward.

iropolo commented on apps/dialtone-documentation/docs/.vuepress/baseComponents/AllTokens.vue:
```diff
@@ -14,50 +52,243 @@
+    <token-table :category="category" :tokens="filteredTokens[category]" />
+  </div>
+  <div
+    v-if="!hasSearchResults"
```
Here also we are using `hasSearchResults` just once and with a `!`.
I will suggest to update the name to sth like:

`noSearchResults`
`const noSearchResults = computed(() => Object.keys(filteredTokens.value).length === 0);`

iropolo commented on apps/dialtone-documentation/docs/.vuepress/baseComponents/AllTokens.vue:
```diff
@@ -14,50 +52,243 @@
 
-<script>
+<script setup>
+import * as tokensJson from '@dialpad/dialtone-tokens/dist/doc.json';
```
`import tokensJson from '@dialpad/dialtone-tokens/dist/doc.json';`
This should be the same.

iropolo commented on apps/dialtone-documentation/docs/.vuepress/baseComponents/AllTokens.vue:
```diff
@@ -14,50 +52,243 @@
+<script setup>
+import * as tokensJson from '@dialpad/dialtone-tokens/dist/doc.json';
+import { getComposedTypographyTokens, getComposedShadowTokens } from '../common/token-utilities';
 import TokenTable, { CATEGORY_MAP, FORMAT_MAP, THEMES } from '@baseComponents/TokenTable.vue';
```
Huh, my idle is blaming a little bit this import. 
`import TokenTable, { CATEGORY_MAP, FORMAT_MAP, THEMES } from '../baseComponents/TokenTable.vue';`
it works well like this and I can go with the definitions.

Beyond that, those exported constant should be separated from the file. 
There is one `apps/dialtone-documentation/docs/.vuepress/common/constants.js`.
But definetly not part of the scope of the ticket.

iropolo commented on apps/dialtone-documentation/docs/tokens/index.md:
```diff
@@ -3,52 +3,11 @@ title: Tokens
-
 Design tokens are all the defined values that are part of the design system, such as color, spacing, typography, and more. Design tokens are published in many different forms so they can be consumed on all platforms (CSS, LESS, Android, iOS, etc).
 
 See the [dialtone-tokens repository](https://github.com/dialpad/dialtone-tokens) for usage info.
```
mmm.. should be this I guess `https://github.com/dialpad/dialtone/pkgs/npm/dialtone-tokens` or.. directly to `https://github.com/dialpad/dialtone` monorepo.

braddialpad replied:
The subrepo for tokens still has it's own readme

https://github.com/dialpad/dialtone/tree/staging/packages/dialtone-tokens

iropolo commented on apps/dialtone-documentation/docs/.vuepress/theme/components/Page.vue:
```diff
@@ -101,4 +102,9 @@ const includeToc = computed(() => {
   return headers?.length > 0;
 });
+
+const tokensPageClass = computed(() => {
+  if (route.path.includes('tokens')) return 'tokens-page';
+  return '';
```
Im guessing this will be escalable soon.

We will think to have this dynamically sth like `pageDynamicallyClasses`
Then it will handle these classes.

ninamarina replied:
We can think about that in the case that some other page needs a custom layout like this one

iropolo reviewed (approved):
Looks good, works well, happy to have this integration!!, nice job

ninamarina committed:
fix for cr suggestions

ninamarina commented:
Thanks @iropolo, just implemented your suggestions
